include/swift/Sema/ConstraintSystem.h

```diff
@@ -658,11 +658,19 @@ struct SelectedOverload {
   /// we're referencing a member.
   const Type openedFullType;
 
+  /// The opened type of the base of the reference to this overload, adjusted
+  /// for `@preconcurrency` or other contextual type-altering attributes.
+  const Type adjustedOpenedFullType;
```

Comment:
What do you think about the idea of doing the opposite here - set `open{Full}Type` to `adjusted` from `DeclReferenceType` and hide "pre-adjustment" types as private members only accessible to `ExprRewriter`? I think it would be safer choice for users like IDE, diagnostics, and concurrency/effect checkers which only need to see the final adjusted type.